src/neml2/solvers/Newton.cxx

```diff
@@ -0,0 +1,128 @@
+// Copyright 2023, UChicago Argonne, LLC
+// All Rights Reserved
+// Software Name: NEML2 -- the New Engineering material Model Library, version 2
+// By: Argonne National Laboratory
+// OPEN SOURCE LICENSE (MIT)
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "neml2/solvers/Newton.h"
+#include <iomanip>
+#include "neml2/misc/math.h"
+
+namespace neml2
+{
+register_NEML2_object(Newton);
+
+OptionSet
+Newton::expected_options()
+{
+  OptionSet options = NonlinearSolver::expected_options();
+  return options;
+}
+
+Newton::Newton(const OptionSet & options)
+  : NonlinearSolver(options)
+{
+}
+
+std::tuple<bool, size_t>
+Newton::solve(NonlinearSystem & system, BatchTensor & x)
+{
+  neml_assert_dbg(!x.requires_grad(), "The trial solution shall not contain any function graph.");
+
+  // The initial residual for relative convergence check
+  system.residual();
+  auto nR0 = system.residual_norm();
+
+  // Check for initial convergence
+  if (converged(0, nR0, nR0))
+  {
+    // FIXME: The final update is only necessary if we use AD
```

Comment:
Do you really mean FIXME?  Or do you mean FIXME in the sense we should only do this if we're set to AD?

Reply:
This probably should not be a TODO rather than FIXME.